test/testpushnotifications.cpp

```diff
@@ -0,0 +1,351 @@
+        const auto socket = processTextMessageSpy.at(0).at(0).value<QWebSocket *>();
+        socket->sendTextMessage("notify_file");
+
+        // filesChanged signal should be emitted
+        QVERIFY(filesChangedSpy.wait());
+        QCOMPARE(filesChangedSpy.count(), 1);
+        // FIXME: This fails because accountFilesChanged is always nullptr. What do I wrong?
+        // I have checked that in a normal signal handler I get the correct account pointer.
+        // auto accountFilesChanged = filesChangedSpy.at(0).at(0).value<OCC::Account *>();
+        // QCOMPARE(accountFilesChanged, account);
+    }
+
+    void testOnWebSocketTextMessageReceived_invalidCredentialsMessage_reconnectWebSocket()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+        QSignalSpy spy(&mockServer, &MockWebSocketServer::processTextMessage);
+        QVERIFY(spy.isValid());
+
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+
+        // Setup push notifications
+        pushNotifications->setup();
+
+        // Wait for authentication attempt and then sent invalid credentials
+        spy.wait();
+        QCOMPARE(spy.count(), 2);
+        const auto socket = spy.at(0).at(0).value<QWebSocket *>();
+        const auto firstPasswordSent = spy.at(1).at(1).toString();
+        QCOMPARE(firstPasswordSent, password);
+        socket->sendTextMessage("err: Invalid credentials");
+
+        // Wait for a new authentication attempt
+        spy.wait();
+        QCOMPARE(spy.count(), 4);
+        const auto secondPasswordSent = spy.at(3).at(1).toString();
+        QCOMPARE(secondPasswordSent, password);
+    }
+
+    void testOnWebSocketError_connectionLost_emitConnectionLost()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+        QSignalSpy processTextMessageSpy(&mockServer, &MockWebSocketServer::processTextMessage);
+        QVERIFY(processTextMessageSpy.isValid());
+
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+        QSignalSpy connectionLostSpy(pushNotifications.get(), &OCC::PushNotifications::connectionLost);
+        QVERIFY(connectionLostSpy.isValid());
+
+        // Setup push notifications
+        pushNotifications->setup();
+
+        // Wait for authentication and then sent a network error
+        processTextMessageSpy.wait();
+        QCOMPARE(processTextMessageSpy.count(), 2);
+        // FIXME: This a little bit ugly but I had no better idea how to trigger a error on the websocket client.
```

Comment:
Well, you're controlling the server now, you could always get it to forcibly terminate the client connections.

Reply:
Have you any idea how to do that? I didn't find a method or way so that the error exception on the client side will be raised?

Reply:
Well, you could close the server side websocket, there are plenty of different ways to close that one. You can also simulate other errors by setting a very small value for maxAllowedIncomingFrameSize etc.

Reply:
abort() was the function to use. Now it works:)